### DISCUSSION TITLE
Running Prairielearn in production

### DISCUSSION THREAD

**Discussion by Prostigus** [PrairieLearn/PrairieLearn#6019] [Q&A] [answered]
Theoretically, if we were to run prairielearn in a production scenario, we would need the have users in the system. 

I know the system assigns one dev user when we start it, but would it be possible to add in instructor or student users that are locked to their own view? (i.e. we remove all things associated with the dev view for them?)

**Comment by nwalters512** [comment, marked as answer]
In production environments, you'd want to run with `NODE_ENV=production`, which would avoid creating the dev user.

For authentication, there are a few options:

- If you want to use Shibboleth, you can run PL behind a reverse proxy with Shib support. Check out `pages/authCallbackShib/authCallbackShib.js` to get a sense of what headers you'd need your reverse proxy to pass.

- You can also use Azure or Google OAuth 2 for authentication. Check out the corresponding `authLogin*` and `authCallback*` pages in the `pages/` directory to get a sense of how to configure them.

- You could do everything through LTI 1.1. This is not currently recommended, as the user experience isn't as good as with full authentication. We're exploring LTI 1.3 which should improve things, but that's not ready for use yet.

**Reply by Prostigus** [reply]
Thank you for your help, I'll give this a go!

**Reply by Prostigus** [reply]
I've given a go at creating my own course and some elements, and now am trying to view it through the production environment. 
I've added the exampleCourse and my custom course to the config.json and copy that over to the docker container. 

I don't see either courses available to sign up for in the production view. I imagine I am doing something wrong and need a bit of guidance to solve this problem.

![image](https://user-images.githubusercontent.com/71531356/179441457-92d61f58-dcff-4206-8793-c30957d87c00.png)
![image](https://user-images.githubusercontent.com/71531356/179441482-161169c6-5635-4577-848d-788c57ec7767.png)

**Reply by nwalters512** [reply]
Does your config file reflect the path of the course inside the container?

**Reply by Prostigus** [reply]
Yes, the files load properly in the Dev environment
![image](https://user-images.githubusercontent.com/71531356/179442283-ef58ce74-ef13-4648-b366-04242c4ca6f8.png)
![image](https://user-images.githubusercontent.com/71531356/179442319-521f793d-4849-495b-9464-3f42fb9712bd.png)

**Reply by nwalters512** [reply]
Did you configure the access rules for your course to allow access? See "Course instance allowAccess" on https://prairielearn.readthedocs.io/en/latest/courseInstance/.

**Reply by Prostigus** [reply]
Yes, I've made sure the allowAccess rules are setup correctly,
![image](https://user-images.githubusercontent.com/71531356/179448640-6d45bd06-0717-4796-83ad-aa4d7bb2913f.png)

**Reply by nwalters512** [reply]
Are you an instructor for the course? Instructors can't enroll in their own courses. Also note that you can't enroll in the example course. See the following SQL query to see how courses are selected for display on the enrollment page: https://github.com/PrairieLearn/PrairieLearn/blob/693ccd8cde2a9353698b083f724ba5df5aa559da/pages/enroll/enroll.sql#L20

**Reply by Prostigus** [reply]
Is there a way to create a user for testing for use in the production environment without setting up Shibboleth or other authentication first?

**Reply by nwalters512** [reply]
Check out the `select_or_insert_user` admin query, available at `/pl/administrator/query/select_or_insert_user`.

**Reply by Prostigus** [reply]
Ok I've almost got Google Oauth2 in place, but in the `"googleRedirectUrl"` parameter, am I supposed to place the page for the callback? I.E. `.../pl/oauth2callback`

**Reply by nwalters512** [reply]
Exactly - e.g. on prairielearn.org, it's set to https://www.prairielearn.org/pl/oauth2callback.

**Reply by evanderpoel** [reply]
To go back to getting the course into our production deployment. We currently got Google Oauth2 functioning but when we sign in we want to get one of courses that are in our src files and are contained in our docker container to be loaded for all users to access. Obviously this worked in the dev environment with the "load from disk" option. But with this production environment how will this course appear without this feature. We have completed all that was said above.

**Reply by nwalters512** [reply]
If you haven't already, you should make yourselves admins - see the `administrators` table in the database. You can then visit the admin page and use the "Add course" button. You can enter the appropriate information, including your course's Git repository. You'll need to ensure that `git` is configured with appropriate authentication to pull your course repository.

**Reply by Prostigus** [reply]
Mostly everything is in place now, for the git config do we need to specify an SSH key? Or is there somewhere I specify my git credentials?

**Reply by nwalters512** [reply]
PrairieLearn is agnostic to the method of authentication you use. It expects that you will have configured `git` authentication somehow, but it doesn't need to know or care about how. The only configuration you can make in PL is `gitSshCommand`, which will be provided to `git` via the standard `GIT_SSH_COMMAND` environment variable:

https://github.com/PrairieLearn/PrairieLearn/blob/693ccd8cde2a9353698b083f724ba5df5aa559da/lib/config.js#L89

**Reply by Prostigus** [reply]
I've got it up and running with the course in a GitHub repository. I never got the courses to show up through the docker container in the production environment, but maintaining the GitHub repository is more scalable in the future so I guess this will be fine. Thanks for all your help!

**Reply by firasm** [reply]
wow - this thread was great! I'm so glad this is archived so I can come back and follow some of these steps.

@Prostigus let's get in touch, I think we're doing some very similar stuff and could give each other a hand.
